File: remote/src/main/java/com/kelsos/mbrc/interactors/QueueInteractor.java
package com.kelsos.mbrc.interactors;

import com.google.inject.Inject;
import com.kelsos.mbrc.annotations.MetaDataType;
import com.kelsos.mbrc.annotations.Queue;
import com.kelsos.mbrc.constants.Code;
import com.kelsos.mbrc.dto.requests.NowPlayingQueueRequest;
import com.kelsos.mbrc.services.api.NowPlayingService;
import rx.Observable;
import rx.schedulers.Schedulers;

public class QueueInteractor {
  @Inject private NowPlayingService service;

  public Observable<Boolean> execute(@MetaDataType.Type String type, @Queue.Action String action, int id) {
    final NowPlayingQueueRequest body = new NowPlayingQueueRequest();
    body.setAction(action);
    body.setId(id);
    body.setType(type);
    return service.nowplayingQueue(body)
        .flatMap(baseResponse -> Observable.just(baseResponse.getCode() == Code.SUCCESS))
        .subscribeOn(Schedulers.io());
  }

  public Observable<Boolean> execute(@Queue.Action String action, String path) {
    final NowPlayingQueueRequest body = new NowPlayingQueueRequest();
    body.setAction(action);
    body.setPath(path);
    return service.nowplayingQueue(body)
        .flatMap(baseResponse -> Observable.just(baseResponse.getCode() == Code.SUCCESS))
        .subscribeOn(Schedulers.io());
  }
}

File: mbrc-lib/src/main/kotlin/com/kelsos/mbrc/interactors/QueueInteractor.kt
package com.kelsos.mbrc.interactors

import com.google.inject.Inject
import com.kelsos.mbrc.annotations.MetaDataType
import com.kelsos.mbrc.annotations.Queue
import com.kelsos.mbrc.constants.Code
import com.kelsos.mbrc.dto.requests.NowPlayingQueueRequest
import com.kelsos.mbrc.services.api.NowPlayingService
import rx.Observable
import rx.functions.Func1
import rx.schedulers.Schedulers

class QueueInteractor {
    @Inject private lateinit var service: NowPlayingService

    fun execute(@MetaDataType.Type type: String, @Queue.Action action: String, id: Long): Observable<Boolean> {
        val body = NowPlayingQueueRequest()
        body.action = action
        body.id = id
        body.type = type
        return service.nowplayingQueue(body)
                .flatMap<Boolean>(Func1{ Observable.just(it.code == Code.SUCCESS) })
                .subscribeOn(Schedulers.io())
    }

    fun execute(@Queue.Action action: String, path: String): Observable<Boolean> {
        val body = NowPlayingQueueRequest()
        body.action = action
        body.path = path
        return service.nowplayingQueue(body)
                .flatMap<Boolean>(Func1{ Observable.just(it.code == Code.SUCCESS) })
                .subscribeOn(Schedulers.io())
    }
}
